drivers/sensor/microchip/mtch9010/mtch9010.c

```diff
@@ -0,0 +1,792 @@
+					     "single sample fallback");
+				mtch9010_command_send(uartDev, config, "0");
+
+				/* Set to 0, since we can't trust the results */
+				devData->reference = 0;
+			}
+#endif
+		} else if (config->refMode == MTCH9010_REFERENCE_CUSTOM_VALUE) {
+			/* The user manually set a value */
+
+			if (devData->reference > MTCH9010_MAX_RESULT) {
+				LOG_INST_ERR(config->log, "Reference out of range");
+				return -EIO;
+			}
+
+			mtch9010_command_send(uartDev, config, "2");
+			sprintf(tBuffer, "%u", devData->reference);
+			mtch9010_command_send(uartDev, config, tBuffer);
+		}
+
+		if (devData->threshold > MTCH9010_MAX_RESULT) {
+			LOG_INST_ERR(config->log, "Threshold out of range");
+			return -EINVAL;
+		}
+
+		/* Detection Threshold */
+		sprintf(tBuffer, "%u", devData->threshold);
+		mtch9010_command_send(uartDev, config, tBuffer);
+
+#ifdef CONFIG_MTCH9010_LOCK_ON_STARTUP
+		mtch9010_lock_device(dev);
+#endif
+	} else {
+		LOG_INST_DBG(config->log, "UART setup not enabled");
+
+#ifdef CONFIG_MTCH9010_LOCK_ON_STARTUP
+		BUILD_ASSERT(true, "MTCH9010_LOCK_ON_STARTUP cannot be set without UART enabled");
+#endif
+	}
+
+#ifdef CONFIG_MTCH9010_OVERRIDE_DELAY_ENABLE
+	/* Print warning */
+	if (config->sleepTime != 0) {
+		uint16_t timeout =
+			(config->sleepTime) * 1000 + CONFIG_MTCH9010_SAMPLE_DELAY_TIMEOUT_MS;
+		LOG_INST_WRN(config->log, "Device will wait up-to %u ms when fetching samples",
+			     (timeout));
+	}
+#endif
+
+	LOG_INST_DBG(config->log, "MTCH9010 configuration complete");
+
+	return ret;
+}
+
+static int mtch9010_command_send(const struct device *dev, const struct mtch9010_config_t *config,
+				 const char *str)
+{
+#ifdef CONFIG_MTCH9010_ECHO_UART_TX_ENABLED
+	LOG_INST_INF(config->log, "\"%s\"", str);
+#endif
+
+	uint8_t len = strlen(str);
+
+	for (uint8_t i = 0; i < len; i++) {
+		uart_poll_out(dev, str[i]);
+	}
+
+	uart_poll_out(dev, MTCH9010_SUBMIT_CHAR);
+
+	char rxBuffer[4] = {'\0', '\0', '\0', '\0'};
+
+	if (mtch9010_timeout_receive(dev, &rxBuffer[0], 3,
+				     CONFIG_MTCH9010_UART_COMMAND_TIMEOUT_MS) == 0) {
+		LOG_INST_ERR(config->log, "CMD Timeout waiting for response");
+		return -EIO;
+	}
+
+	if (rxBuffer[0] == MTCH9010_ACK_CHAR) {
+#ifdef CONFIG_MTCH9010_ECHO_UART_TX_ENABLED
+		LOG_INST_DBG(config->log, "ACK received");
+#endif
+		return 0;
+	} else if (rxBuffer[0] == MTCH9010_NACK_CHAR) {
+		LOG_INST_ERR(config->log, "NACK received from command");
+		return -EIO;
+	}
+	LOG_INST_ERR(config->log, "Unknown character received during setup");
+	return -EIO;
+}
+
+/* Returns the number of bytes received */
+static int mtch9010_timeout_receive(const struct device *dev, char *buffer, uint8_t maxLen,
+				    uint16_t milliseconds)
+{
+	uint8_t count = 0;
+	k_timepoint_t end = sys_timepoint_calc(K_MSEC(milliseconds));
+
+	while ((count < maxLen) && (!sys_timepoint_expired(end))) {
+		int code = uart_poll_in(dev, &buffer[count]);
+
+		if (code == 0) {
+			count++;
+		}
+	}
```

Comment:
Do responses end with `\r\n`? If yes - we can terminate reception when `\r\n` is received this making reception faster.